vivisect/analysis/i386/golang.py

```diff
@@ -0,0 +1,202 @@
+'''
+Perform further analysis on 32-bit GO language executables.
+
+This module extends code flow analysis to runtime_main for Windows PE binaries.
+GO binaries start from a single export and proceed thru several functions
+that initialize GO.  Specific application code is launched from the GO function
+runtime_main(), which is invoked by "call eax" with an address placed on the
+stack many calls earlier.  Vivisect code flow analysis does not track the
+address; this module finds the address and invokes makeFunction(va).
+
+Samples have been found with different instruction sequences for reaching the
+address of runtime_main(); this module attempts all observed sequences.
+'''
+
+import envi
+import envi.archs.i386.disasm
+
+
+def analyze(vw):
+    '''
+    Perform further analysis on GO language executables.
+    '''
+    # Make sure it is a PE file, with "Go build ID:" in the first few bytes
+    # of the .text segment (versus a .upxN segment of a packed sample).
+    if (vw.getMeta('Format') != 'pe') and \
+       (vw.getMeta('Architecture') != 'i386'):
+        return
+    has_go_build = False
+    for segment in vw.getSegments():
+        va, size, name, filename = segment
+        if name == '.text':
+            bytez = vw.readMemory(va, 10000)
+            k1 = bytez.find(b'Go build ID: ')
+            if k1 != -1:
+                has_go_build = True
+                break
+    if not has_go_build:
+        return
+
+    # There should be just one entry point (the public export).
+    ep = vw.getEntryPoints()
+    if len(ep) != 1:
+        return
+
+    # Search from the entry point function.
+    ep_va = ep[0]
+    bblocks = vw.getFunctionBlocks(ep_va)
+    if not bblocks:
+        return
+    ptr_va, runtime_va = extract_golang_mainmain(vw, bblocks, filename)
+    if runtime_va is None:
+        return
```

Comment:
That's because we don't see any.  The modules were developed to detect runtime_main over  a set of 10,000 GO samples obtained from-the-wild (no easy cases).  The methods in the code are successful 94% of the time.

Reply:
cool!  that's a nice test-set size, assuming the set covers a wide range of code-variety (knowing where you are most likely to have obtained then, i'm guessing that's a yes).

point of interest: what kinds of things make up the 6% which are not successful?

Reply:
I looked at a couple in detail and they were just weird.  Our test set probably covers many versions of GO.  I'm happy with 94% as a first effort.